pkg/collector/corechecks/cluster/helm/releases_store.go

```diff
@@ -0,0 +1,89 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build kubeapiserver
+// +build kubeapiserver
+
+package helm
+
+import "sync"
+
+type releasesStore struct {
+	store      map[helmStorage]map[namespacedName]map[revision]*release
+	storeMutex sync.Mutex
+}
+
+func newReleasesStore() *releasesStore {
+	res := &releasesStore{
+		store: make(map[helmStorage]map[namespacedName]map[revision]*release),
+	}
+
+	for _, storageDriver := range []helmStorage{k8sConfigmaps, k8sSecrets} {
+		res.store[storageDriver] = make(map[namespacedName]map[revision]*release)
+	}
+
+	return res
+}
+
+// add stores a release
+func (rs *releasesStore) add(rel *release, storage helmStorage) {
+	rs.storeMutex.Lock()
+	defer rs.storeMutex.Unlock()
+
+	if rs.store[storage][rel.namespacedName()] == nil {
+		rs.store[storage][rel.namespacedName()] = make(map[revision]*release)
+	}
+
+	rs.store[storage][rel.namespacedName()][rel.revision()] = rel
+}
+
+// get returns the release stored with the given namespacedName, revision and
+// storage. Returns nil when it does not exist.
+func (rs *releasesStore) get(namespacedName namespacedName, revision revision, storage helmStorage) *release {
+	rs.storeMutex.Lock()
+	defer rs.storeMutex.Unlock()
+
+	if rs.store[storage][namespacedName] == nil {
+		return nil
+	}
+
+	return rs.store[storage][namespacedName][revision]
+}
+
+// getAll returns all the releases stored for the given helmStorage
+func (rs *releasesStore) getAll(storage helmStorage) []*release {
+	rs.storeMutex.Lock()
+	defer rs.storeMutex.Unlock()
+
+	var res []*release
+
+	for _, releasesByRevision := range rs.store[storage] {
+		for _, rel := range releasesByRevision {
+			res = append(res, rel)
+		}
+	}
+
+	return res
+}
+
+// delete removes a release. It returns a bool that indicates whether the
+// release deleted was the only existing revision.
+func (rs *releasesStore) delete(rel *release, storage helmStorage) bool {
+	rs.storeMutex.Lock()
+	defer rs.storeMutex.Unlock()
+
+	if rs.store[storage][rel.namespacedName()] == nil {
+		return true
```

Comment:
This is the nittiest of the nitpicks, but this does not match the comment on this func: no release was deleted because it didn't exist. Perhaps the docs should be rephrased as "it returns a bool that indicates whether there are any revisions left for the release".

EDIT: after re-reading https://github.com/DataDog/datadog-agent/pull/11317/files#diff-a91a2bb9f4306899a82675022adbdb60dcdec035844e948284139007fc4dcb83R342, maybe returning true here might cause multiple delete events. Not sure if it's useful to address it here (it might be a race somewhere else, or never even happen at all), so just throwing the thought out there, I don't exactly have a concern.

Reply:
You're right, the description is not accurate. I'll change this.